handlers/compute.go

```diff
@@ -1,45 +1,142 @@
 package handlers
 
 import (
+	"encoding/json"
 	"github.com/Sirupsen/logrus"
-	"github.com/rancher/agent/handlers/dockerClient"
+	"github.com/mitchellh/mapstructure"
+	"github.com/rancher/agent/handlers/docker"
 	"github.com/rancher/agent/handlers/progress"
 	"github.com/rancher/agent/handlers/utils"
 	"github.com/rancher/agent/model"
 	revents "github.com/rancher/go-machine-service/events"
 	"github.com/rancher/go-rancher/client"
-	"sync"
 )
 
-type InstanceWithLock struct {
-	mu sync.Mutex
-	in *model.Instance
-}
-
 func InstanceActivate(event *revents.Event, cli *client.RancherClient) error {
 	logrus.Infof("Received event: Name: %s, Event Id: %s, Resource Id: %s", event.Name, event.ID, event.ResourceID)
 	instance, host := utils.GetInstanceAndHost(event)
 
-	progress := progress.Progress{}
+	progress := progress.Progress{Request: event, Client: cli}
 
-	if instance != nil {
-		processData, ok := event.Data["processData"]
-		if ok {
-			instance.ProcessData = processData.(map[string]interface{})
-		}
+	if processData, ok := event.Data["processData"]; ok && instance != nil {
+		instance.ProcessData = processData.(map[string]interface{})
+	}
+	if utils.IsNoOp(event.Data) {
+		utils.RecordState(docker.GetClient(utils.DefaultVersion), instance, "")
+		return reply(utils.GetResponseData(event), event, cli)
 	}
 
-	insWithLock := InstanceWithLock{mu: sync.Mutex{}, in: instance}
-	insWithLock.mu.Lock()
-	defer insWithLock.mu.Unlock()
-	if utils.IsInstanceActive(insWithLock.in, host) {
+	if utils.IsInstanceActive(instance, host) {
 		logrus.Info("instance is activated")
-		utils.RecordState(dockerClient.GetClient(utils.DefaultVersion), instance, "")
-		return reply(event.Data, utils.GetResponseData(event, event.Data), cli)
+		utils.RecordState(docker.GetClient(utils.DefaultVersion), instance, "")
+		return reply(utils.GetResponseData(event), event, cli)
 	}
 
 	utils.DoInstanceActivate(instance, host, &progress)
```

Comment:
It doesn't make much sense to me that DoInstanceActivate is in a utils package. It isn't a re-used utility function. It is only used in this specific context. What was your motivation for putting it in the utils package?